File: frontend/src/mq/kafka/cluster.ts
/**
 * Kafka's view of the canonical cluster models.
 *
 * The keys are a contract with internal/driver/kafka/cluster.go.
 *
 * What is absent matters as much as what is here. Kafka's admin protocol
 * reports no rate of any kind - produce and consume rates live in JMX, which
 * this app does not speak - so the canonical rate fields carry the unknown
 * sentinel and no Kafka board draws a per-second figure. Metadata carries no
 * disk figure either; that arrives with the log directories.
 *
 * What Kafka has instead is partition health, and it is the whole reason its
 * overview exists: a partition can be under-replicated, offline, or without a
 * leader entirely, and those are three different degrees of trouble.
 */
import type { ClusterOverview, Node } from "@bindings/model/models";

const AttrNodeID = "nodeId";
const AttrRack = "rack";
const AttrController = "controller";

const AttrClusterID = "clusterId";
const AttrControllerNode = "controllerNode";
const AttrTopicCount = "topics";
const AttrInternalTopicCount = "internalTopics";
const AttrPartitionCount = "partitions";
const AttrUnderReplicated = "underReplicatedPartitions";
const AttrOfflinePartitions = "offlinePartitions";
const AttrLeaderlessPartitions = "leaderlessPartitions";
const AttrGroupCount = "consumerGroups";

function attr(source: { attributes?: Record<string, string | undefined> }, key: string): string {
  return source.attributes?.[key] ?? "";
}

/**
 * A counter the driver did not report reads as unknown, not as zero.
 *
 * The difference is the point: "no partition is under-replicated" and "nobody
 * asked" look identical once both are rendered as 0, and only one of them
 * means the cluster is healthy.
 */
function count(source: { attributes?: Record<string, string | undefined> }, key: string): number | null {
  const raw = attr(source, key);
  if (raw === "") return null;
  const value = Number.parseInt(raw, 10);
  return Number.isNaN(value) ? null : value;
}

export const nodeID = (node: Node): string => attr(node, AttrNodeID);
export const rack = (node: Node): string => attr(node, AttrRack);
export const isController = (node: Node): boolean => attr(node, AttrController) === "true";

export const clusterID = (overview: ClusterOverview): string => attr(overview, AttrClusterID);
export const controllerNode = (overview: ClusterOverview): string =>
  attr(overview, AttrControllerNode);
export const topicCount = (overview: ClusterOverview): number | null =>
  count(overview, AttrTopicCount);
export const internalTopicCount = (overview: ClusterOverview): number | null =>
  count(overview, AttrInternalTopicCount);
export const partitionCount = (overview: ClusterOverview): number | null =>
  count(overview, AttrPartitionCount);
export const underReplicatedPartitions = (overview: ClusterOverview): number | null =>
  count(overview, AttrUnderReplicated);
export const offlinePartitions = (overview: ClusterOverview): number | null =>
  count(overview, AttrOfflinePartitions);
export const leaderlessPartitions = (overview: ClusterOverview): number | null =>
  count(overview, AttrLeaderlessPartitions);
export const consumerGroupCount = (overview: ClusterOverview): number | null =>
  count(overview, AttrGroupCount);

/**
 * Whether anything is wrong with the partitions.
 *
 * One question rather than three because that is how the page is read: an
 * operator glances at the overview to find out whether to keep looking, and
 * any one of the three failures means yes.
 */
export function partitionsAreHealthy(overview: ClusterOverview): boolean {
  return [underReplicatedPartitions, offlinePartitions, leaderlessPartitions].every(
    (read) => read(overview) === 0,
  );
}

/*
 * A transaction's age, as a label.
 *
 * Unknown when the coordinator did not report a start time, and unknown again
 * when the answer would be negative: the broker's clock and this machine's are
 * not the same clock, and a transaction that appears to start in the future is
 * skew rather than news.
 */
export function transactionAge(startedAt: number, now: number): string {
  if (startedAt < 0 || now < startedAt) return "—";
  const seconds = Math.floor((now - startedAt) / 1000);
  if (seconds < 60) return `${seconds}s`;
  const minutes = Math.floor(seconds / 60);
  if (minutes < 60) return `${minutes}m`;
  const hours = Math.floor(minutes / 60);
  if (hours < 24) return `${hours}h ${minutes % 60}m`;
  return `${Math.floor(hours / 24)}d ${hours % 24}h`;
}

/**
 * Whether a transaction has outlived the deadline the cluster set for it.
 *
 * This is the one that means something is wrong rather than merely slow: the
 * coordinator undertook to abort the transaction after its timeout, so one
 * still open well past it is not a long-running job, it is a transaction
 * nothing is finishing.
 */
export function transactionOverdue(
  startedAt: number,
  timeoutMs: number,
  now: number,
): boolean {
  if (startedAt < 0 || timeoutMs <= 0) return false;
  return now - startedAt > timeoutMs;
}

File: internal/driver/kafka/cluster.go
package kafka

import (
	"context"
	"fmt"
	"net"
	"sort"
	"strconv"

	"github.com/twmb/franz-go/pkg/kadm"

	"github.com/amigoer/mq-studio/internal/model"
)

/*
 * Attribute keys the cluster and overview boards read.
 *
 * A private contract between this driver and frontend/src/mq/kafka/cluster.ts,
 * not part of the shared vocabulary: what a Kafka broker reports about itself
 * is not what a RocketMQ or RabbitMQ one does.
 */
const (
	// Per node.
	AttrNodeID     = "nodeId"
	AttrRack       = "rack"
	AttrController = "controller"

	// Per cluster.
	AttrClusterID           = "clusterId"
	AttrControllerNode      = "controllerNode"
	AttrBrokerCount         = "brokers"
	AttrTopicCount          = "topics"
	AttrInternalTopicCount  = "internalTopics"
	AttrPartitionCount      = "partitions"
	AttrUnderReplicated     = "underReplicatedPartitions"
	AttrOfflinePartitions   = "offlinePartitions"
	AttrLeaderlessPartition = "leaderlessPartitions"
	AttrGroupCount          = "consumerGroups"
)

// ListNodes reports the brokers the cluster is made of.
func (c *Conn) ListNodes(ctx context.Context) ([]*model.Node, error) {
	metadata, err := c.admin.BrokerMetadata(fresh(ctx))
	if err != nil {
		return nil, err
	}
	return nodesFrom(metadata), nil
}

// NodeDetail reports one broker, found by the address ListNodes gave it.
func (c *Conn) NodeDetail(ctx context.Context, address string) (*model.Node, error) {
	nodes, err := c.ListNodes(ctx)
	if err != nil {
		return nil, err
	}
	for _, node := range nodes {
		if node.Address == address || node.Name == address {
			return node, nil
		}
	}
	return nil, fmt.Errorf("broker not found: %s", address)
}

// ClusterOverview is the header the cluster and overview pages share.
//
// Kafka has no running totals to ask for - there is no endpoint that answers
// "how many partitions are under-replicated". The figures come from walking
// the metadata of every topic, which is one request, and that walk is the only
// place the health of the cluster exists at all.
func (c *Conn) ClusterOverview(ctx context.Context) (*model.ClusterOverview, error) {
	metadata, err := c.admin.Metadata(fresh(ctx))
	if err != nil {
		return nil, err
	}
	health := healthOf(metadata)

	overview := &model.ClusterOverview{
		Name:        metadata.Cluster,
		TotalNodes:  len(metadata.Brokers),
		OnlineNodes: len(metadata.Brokers),
		// Internal topics are excluded: __consumer_offsets is not something a
		// user made, and counting it makes an empty cluster look populated.
		Destinations: health.topics,
		// Kafka reports no disk figure in metadata. It is in the log dir
		// listing, which is a request per broker and does not belong here.
		AvgDiskUsage:  model.UnknownMetric,
		Subscriptions: model.UnknownMetric,
		Attributes: map[string]string{
			AttrClusterID:           metadata.Cluster,
			AttrControllerNode:      controllerName(metadata),
			AttrBrokerCount:         strconv.Itoa(len(metadata.Brokers)),
			AttrTopicCount:          strconv.Itoa(health.topics),
			AttrInternalTopicCount:  strconv.Itoa(health.internalTopics),
			AttrPartitionCount:      strconv.Itoa(health.partitions),
			AttrUnderReplicated:     strconv.Itoa(health.underReplicated),
			AttrOfflinePartitions:   strconv.Itoa(health.offline),
			AttrLeaderlessPartition: strconv.Itoa(health.leaderless),
		},
	}

	// Groups are a second request, and a cluster that refuses to list them
	// still has a usable overview - so the count is left unknown rather than
	// failing the whole page.
	if groups, err := c.admin.ListGroups(ctx); err == nil {
		overview.Subscriptions = len(groups)
		overview.Attributes[AttrGroupCount] = strconv.Itoa(len(groups))
	}
	return overview, nil
}

// clusterHealth is what one metadata walk yields.
type clusterHealth struct {
	topics         int
	internalTopics int
	partitions     int

	// underReplicated is a partition with fewer in-sync replicas than
	// replicas. It is the first sign of a broker falling behind and the
	// reason this page exists.
	underReplicated int
	// offline is a partition with a replica the cluster cannot reach.
	offline int
	// leaderless is a partition with no leader at all, which means it is
	// neither readable nor writable right now.
	leaderless int
}

func healthOf(metadata kadm.Metadata) clusterHealth {
	var health clusterHealth
	for _, topic := range metadata.Topics {
		if topic.IsInternal {
			health.internalTopics++
		} else {
			health.topics++
		}
		for _, partition := range topic.Partitions {
			health.partitions++
			if partition.Leader < 0 {
				health.leaderless++
			}
			if len(partition.OfflineReplicas) > 0 {
				health.offline++
			}
			if len(partition.ISR) < len(partition.Replicas) {
				health.underReplicated++
			}
		}
	}
	return health
}

func nodesFrom(metadata kadm.Metadata) []*model.Node {
	brokers := append(kadm.BrokerDetails(nil), metadata.Brokers...)
	sort.Slice(brokers, func(i, j int) bool { return brokers[i].NodeID < brokers[j].NodeID })

	nodes := make([]*model.Node, 0, len(brokers))
	for _, broker := range brokers {
		nodes = append(nodes, nodeFrom(broker, metadata.Controller))
	}
	return nodes
}

func nodeFrom(broker kadm.BrokerDetail, controller int32) *model.Node {
	rack := ""
	if broker.Rack != nil {
		rack = *broker.Rack
	}
	return &model.Node{
		// The broker's own node id, which is stable and unique across the
		// cluster - a better list key than a position in a slice, and the
		// number every Kafka tool and log line refers to a broker by.
		ID:      int(broker.NodeID),
		Name:    brokerName(broker),
		Address: brokerAddress(broker),
		Cluster: rack,
		Status:  model.NodeOnline,
		// A broker in metadata is a broker that answered. Kafka reports no
		// rate and no disk figure here, and a zero would read as "measured,
		// and it is zero".
		RateIn:    model.UnknownMetric,
		RateOut:   model.UnknownMetric,
		DiskUsage: model.UnknownMetric,
		Attributes: map[string]string{
			AttrNodeID:     strconv.FormatInt(int64(broker.NodeID), 10),
			AttrRack:       rack,
			AttrController: strconv.FormatBool(broker.NodeID == controller),
		},
	}
}

// brokerName is what an operator calls this broker: its node id, which is what
// every Kafka command line, log line and reassignment plan names it by.
func brokerName(broker kadm.BrokerDetail) string {
	return "broker-" + strconv.FormatInt(int64(broker.NodeID), 10)
}

func brokerAddress(broker kadm.BrokerDetail) string {
	return net.JoinHostPort(broker.Host, strconv.FormatInt(int64(broker.Port), 10))
}

// controllerName is the node id of the controller, or empty when the cluster
// does not name one. -1 is what the protocol sends for "no controller", and
// rendering that as a broker id would invent a broker.
func controllerName(metadata kadm.Metadata) string {
	if metadata.Controller < 0 {
		return ""
	}
	return strconv.FormatInt(int64(metadata.Controller), 10)
}
